File: 2019/day5/day5.py
f = open('input', 'r')

inst = f.read().split(',')

def proc(tape, inp):
    pos = 0
    while not tape[pos] == '99':
        if tape[pos][-1] == '1': #add
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            add = int(x) + int(y)
            tape[int(tape[pos+3])] = str(add)
            pos += 4
        elif tape[pos][-1] == '2': #mul
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            prod = int(x) * int(y)
            tape[int(tape[pos+3])] = str(prod)
            pos += 4
        elif tape[pos] == '3': #read
            i = int(tape[pos+1])
            tape[i] = str(inp)
            pos += 2
        elif tape[pos][-1] == '4': #output
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                o = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                o = tape[pos+1]
            print(o)
            pos += 2
        elif tape[pos][-1] == '5': #jtrue
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            if not int(x) == 0:
                pos = int(y)
            else:
                pos += 3
        elif tape[pos][-1] == '6': #jfalse
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            if int(x) == 0:
                pos = int(y)
            else:
                pos += 3
        elif tape[pos][-1] == '7': #lessthan
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            if int(x) < int(y):
                tape[int(tape[pos+3])] = '1'
            else:
                tape[int(tape[pos+3])] = '0'
            pos += 4
        elif tape[pos][-1] == '8': #equals
            if len(tape[pos]) < 3 or tape[pos][-3] == '0':
                x = tape[int(tape[pos+1])]
            elif tape[pos][-3] == '1':
                x = tape[pos+1]
            if len(tape[pos]) < 4 or tape[pos][-4] == '0':
                y = tape[int(tape[pos+2])]
            elif tape[pos][-4] == '1':
                y = tape[pos+2]
            if int(x) == int(y):
                tape[int(tape[pos+3])] = '1'
            else:
                tape[int(tape[pos+3])] = '0'
            pos += 4
        else:
            print('error')
            return -1
    return tape[0]

proc(inst, 5)

File: 2019/day5/day5.cpp
#include <stdio.h>
#include <tuple>
#include <utility>
#include <vector>

std::pair<int, int> getparams(std::vector<int> mem, int pos)
{
    int x, y, inst = mem[pos];

    // get x
    int xmode = (inst / 100) % 10;
    if (xmode == 0) // position
    {
        x = mem[mem[pos + 1]];
    }
    else if (xmode == 1) // immediate
    {
        x = mem[pos + 1];
    }
    else // error
    {
        printf("getparam: x mode error: %d\n", xmode);
        return std::make_pair(-1, -1);
    }

    // only 1 param
    if (inst % 10 == 3 || inst % 10 == 4)
        return std::make_pair(x, 0);

    // get y
    int ymode = (inst / 1000) % 10;
    if (ymode == 0) // position
    {
        y = mem[mem[pos + 2]];
    }
    else if (ymode == 1) // immediate
    {
        y = mem[pos + 2];
    }
    else // error
    {
        printf("getparam: y mode error: %d\n", ymode);
        return std::make_pair(-1, -1);
    }

    return std::make_pair(x, y);
}

bool proc(std::vector<int> mem)
{
    int pos = 0;
    while (mem[pos] != 99)
    {
        int inst = mem[pos];
        switch (inst % 10)
        {
            case 1: // add
            {
                int x, y;
                std::tie(x, y) = getparams(mem, pos);
                mem[mem[pos + 3]] = x + y;
                pos += 4;
                break;
            }
            case 2: // mul
            {
                int x, y;
                std::tie(x, y) = getparams(mem, pos);
                mem[mem[pos + 3]] = x * y;
                pos += 4;
                break;
            }
            case 3: // input
            {
                mem[mem[pos + 1]] = 5;
                pos += 2;
                break;
            }
            case 4: // output
            {
                int o;
                std::tie(o, std::ignore) = getparams(mem, pos);
                printf("%d\n", o);
                pos += 2;
                break;
            }
            case 5: // jtrue
            {
                int x, y;
                std::tie(x, y) = getparams(mem, pos);
                if (x != 0)
                    pos = y;
                else
                    pos += 3;
                break;
            }
            case 6: // jfalse
            {
                int x, y;
                std::tie(x, y) = getparams(mem, pos);
                if (x == 0)
                    pos = y;
                else
                    pos += 3;
                break;
            }
            case 7: // lessthan
            {
                int x, y;
                std::tie(x, y) = getparams(mem, pos);
                if (x < y)
                    mem[mem[pos + 3]] = 1;
                else
                    mem[mem[pos + 3]] = 0;
                pos += 4;
                break;
            }
            case 8: // equals
            {
                int x, y;
                std::tie(x, y) = getparams(mem, pos);
                if (x == y)
                    mem[mem[pos + 3]] = 1;
                else
                    mem[mem[pos + 3]] = 0;
                pos += 4;
                break;
            }
            default:
            {
                printf("instruction error: %d\n", inst);
                return false;
            }
        }
    }

    return true;
}

int main()
{
    std::vector<int> mem;
    for (int i = 0; i < 678; i++)
    {
        int p;
        scanf("%d,", &p);
        mem.push_back(p);
    }

    proc(mem);
}